plugins/synapticsmst/synapticsmst-device.c

```diff
@@ -504,32 +541,502 @@ synapticsmst_device_get_flash_checksum (SynapticsMSTDevice *device,
+{
+	
+	guint32 dwData[16];
+	guint32 fw_size;
+	guint32 checksum = 0;
+	guint32 flash_checksum = 0;
+	gboolean  eraseFail = FALSE;
+	guint8 rc = 0xFF;
+	guint8 BankInUse = BANKTAG_0;
+	guint8 BankToUpdate = BANKTAG_1;
+	guint8 tagData[16], readBuf[256];
+	guint16 tmpCRC =0;	
+	guint32 i = 0;
+	guint32 writeOffset, writeIndex, eraseOffset;
+	guint32 write_loops = 0;
+	guint32 retryCount, unitSize;
+	time_t timeptr;
+	struct tm *pTM;
+	SynapticsMSTDevicePrivate *priv = GET_PRIVATE (device);
+	g_autoptr(SynapticsMSTConnection) connection = NULL;
+
+	connection = synapticsmst_common_new (priv->fd, priv->layer, priv->rad);
+	/* get used bank */
+	rc = 0;
+	rc = synapticsmst_common_rc_get_command (connection, UPDC_READ_FROM_MEMORY,
+						((sizeof(dwData)/sizeof(dwData[0]))*4), (gint)0x20010c, (guint8*)dwData);
+	if(rc)								
+		g_debug("Get Bank fail");		
+
+	if((dwData[0] & BIT(7))||(dwData[0] & BIT(30))){
+		BankInUse = BANKTAG_1;
+		BankToUpdate = BANKTAG_0;
+	}
+	#if MSTDBG
+	g_debug("Used %x ToUpd %x",BankInUse, BankToUpdate);
+	#endif
+	/* get firmware size */
+	fw_size =  0x410 + (*(payload_data + 0x400) << 24) + (*(payload_data + 0x401) << 16) + (*(payload_data + 0x402) << 8) + (*(payload_data + 0x403));
+	if( fw_size < payload_len ){
+		fw_size = 104*1024;		// Current firmware size is 104K
+	}
+
+	rc = 0;
+	retryCount = 0;	
+	unitSize = BLOCK_UNIT;//64
+	write_loops = fw_size / unitSize;//(fw_size / unitSize / 6);//Here write 256 byte?
+	if (fw_size % unitSize)
+		write_loops++;
+
+	do{
+		/* erase storage */
+		eraseOffset = BankToUpdate * 2;
+		rc = synapticsmst_device_set_flash_sector_erase(device, FLASH_SECTOR_ERASE_64K, eraseOffset++);
+		if (rc)
+		{
+			g_set_error_literal (error,	G_IO_ERROR,	G_IO_ERROR_INVALID_DATA,
+						"can't sector erase flash");
+			eraseFail = TRUE;
+			break;
+		}
+		rc = synapticsmst_device_set_flash_sector_erase(device, FLASH_SECTOR_ERASE_64K, eraseOffset);
+		if (rc)
+		{
+			g_set_error_literal (error,	G_IO_ERROR,	G_IO_ERROR_INVALID_DATA,
+						"can't sector erase flash");
+			eraseFail = TRUE;
+			break;
+		}
+		if(eraseFail)
+			break;
+		/* update */
+		#if MSTDBG
+		g_debug("skip check ESM update");
+		#endif
+		writeIndex = 0;	
+		writeOffset = EEPROM_BANK_OFFSET * BankToUpdate;		
+		
+		for( i = 0; i < write_loops ; i++ )
+		{
+			rc = synapticsmst_common_rc_set_command (connection,
+					 UPDC_WRITE_TO_EEPROM, unitSize, writeOffset, payload_data+ writeIndex);
+			if (rc) {
+				/* repeat once */
+				rc = synapticsmst_common_rc_set_command (connection,
+						 UPDC_WRITE_TO_EEPROM, unitSize, writeOffset, payload_data+ writeIndex);
+			}
+			if (rc)	{
+				g_set_error_literal (error,	G_IO_ERROR,	G_IO_ERROR_INVALID_DATA,
+							"FW write fail");
+				break;
+			}
+
+			writeOffset += unitSize;
+			writeIndex += unitSize;
+		}	
+
+		/* verify CRC */
+		rc = 0;
+		checksum = synapticsmst_device_get_crc( 0, 16, fw_size, payload_data );
+		for(i=0; i<5; i++)
+		{
+			g_usleep(2000);	
```

Comment:
Does is actually need to sleep in between CRC checks?  I would think sleep shouldn't be needed here.

Reply:
It just follow win32 application flow, calculate need delay time, maybe could get shorten.

Reply:
If it does need delay between checks due to remote command latency or other reason that's OK just please add a comment to indicate that.